openmls/TEST-DESIGN.md

```diff
@@ -0,0 +1,172 @@
+# OpenMLS testing proposition
+
+The purpose of this document is to improve tests in openmls and try to agree on a common standard for writing them.
+
+## Current status
+
+Here are some factual remarks about the current state of tests.
+
+1. erratic: no clear pattern emerges. It looks like tests have been piled up at different times by different persons.
+   There are very few unit tests, some integration tests in `src/` and some other integration tests in `tests/`
+2. too verbose: some domain are of course complex and hard to test, but some simpler ones require thousands of lines to
+   get tested
+3. not idiomatic: one of the recommended way to write tests in Rust is to write unit tests in the same file as the
+   sources in order to have access to private members. Integration tests (in `tests/`) are more high level and
+   especially
+   used for verifying the public API
+4. Not enough factorization: it is not uncommon to have a file redefining its own set of helpers. A lot of those helpers
+   are duplicated all across the codebase
+5. some tests actually test more than one thing. It makes them harder to write/understand. Also harder to debug when
+   they
+   fail (in which iteration am I ?). This results mostly from the test setup being verbose which incentivizes reuse
+
+## Goals
+
+This proposition aims at:
+
+1. Write less verbose tests and quicker
+2. Share a common pattern for writing tests
+3. Have finer grained tests and be quicker to troubleshoot the cause of a failing test
+4. Be idiomatic to welcome easier external contributions
+
+## Rules
+
+To achieve these goals, here are the "rules"/recommendations to set up:
+
+1. Any test expected to fail by altering a valid input should also keep this valid input and verify that, contrary to
+   the
+   tampered one, it succeeds.
+2. All errors should be namely asserted i.e. `assert!(result.is_err())` is prohibited. When matching an error, one
+   should `assert!(matches!(result.unwrap_err(), MlsError::MyError))` because errors are not supposed to impl Eq (it can
+   impose hard to fulfill constraints on the data they wrap)
+3. Do not use `.expect("...")` but only use `.unwrap()`. The former adds too much overhead for too little value. Another
```

Comment:
In some cases, you can really feel how motivated the creator of a test was about using `expect` instead of `unwrap`: `.expect("An unexpected error occurred.")` 🙈 

So, I generally agree. But let's not phrase it this strict: Does the extra information in `expect` help to debug the problem? If yes, go for it. If no, maybe use `unwrap`. Let's just be sensible about what we use when.

Reply:
There are also different [message styles](https://doc.rust-lang.org/std/error/index.html#common-message-styles) for `expect`.

Reply:
Okay to cool down the rule :)

Reply:
Maybe we should first discuss this before you invest the effort to make changes. The above is only how I think about this and maybe others view it differently :-)